[15721] Data Compression

suafeng opened the pull request:
Add dictionary encoding.

## Created: 
dictionary_encoding_tile.h
dictionary_encoding_tile.cpp

### Description
Create a new class DictEncodedTile. This class inherits from Tile class. The most important functions of DictEncodedTile are DictEncode(), which takes a pointer of a tile as argument and encodes this tile's content, and DictDecode(), which return a new tile with the decoded data.

The main idea of encoding is to replace VARCHAR and VARBINARY data with TINYINT (or other numeric types), the mappings are stored as fields in the class.

## Modified:
tile.h
tile.cpp

### Description
Make minor changes to these files.

## Testcase
tile_compress_test.cpp

### Description

Correctness test. Check whether the original tile is the same after encoding and decoding. Check the correctness of GetValue() function.

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+type::Value DictEncodedTile::GetValue(const oid_t tuple_offset, const oid_t column_id) {
+	if (dict_encoded_columns.count(column_id) > 0) {
+		// encoded
+		auto idx_val = Tile::GetValue(tuple_offset, column_id);
```
Consider renaming this variable. It wasn't immediately clear what idx_val was. Consider changing "idx_val" to "encoded_val" and "idx" to "dict_index".

mbutrovich commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+    }
+	}
+
+  auto *new_tile = new Tile(backend_type, tile_group_header, original_schema, tile_group, num_tuple_slots);
```
Use make_shared instead of new keyword and change return type to std::shared_ptr.

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+	auto curr_val = Tile::GetValueFast(tuple_offset, schema.GetOffset(column_id),
+	  column_type, is_inlined);
+	if (dict_encoded_columns.count(column_id) > 0) {
+		auto idx_val = Tile::GetValue(tuple_offset, column_id);
```
Variable naming again, consider changing "idx_val" to "encoded_val" and "idx" to "dict_index".

mbutrovich commented on src/include/storage/tile.h:
```diff
@@ -13,6 +13,10 @@
 #pragma once
 
 #include <mutex>
+#include <vector>
```
Doesn't look like these includes are needed in tile.h. Can you just put them in dictionary_encoding_tile.h?

mbutrovich commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+	size_t offset = 0;
+
+	// find out which column to encode
+	for (oid_t i = 0; i < column_count; i++) {
```
Recommend more descriptive iterator name like column_idx.

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+							true);
+				}
+				LOG_INFO("%s", curr_val.GetInfo().c_str());
+				// assume the idx take 1 byte
```
Why do you assume that 1 byte is large enough to store the index? What if you have more than 2^8 unique values in this tile's column?

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+				}
+				LOG_INFO("%s", curr_val.GetInfo().c_str());
+				// assume the idx take 1 byte
+				char idx_data[1];
```
Consider renaming this variable. It's not clear what "idx_data" is from the name. Maybe "encoded_indices" ?

mbutrovich commented on test/storage/tile_compression_test.cpp:
```diff
@@ -0,0 +1,181 @@
+  tile->InsertTuple(2, tuple3.get());
+  
+  // encode the tile
+	std::unique_ptr<storage::DictEncodedTile> encodedTile(
```
std::make_unique instead of new -- this applies to pretty much all uses of new in this test.

mbutrovich commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+		if (dict_encoded_columns.count(i) > 0) {
+			LOG_INFO("encoding column %s", schema.GetColumn(i).column_name.c_str());
+			// to for tuple offset
+			for (oid_t to = 0; to < num_tuple_slots; to++) {
```
If this iterator is named tuple_offset you don't need the comment above explaining what to means.

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+	// use a 2d vector to save the data
+	std::vector<std::vector<type::Value>> new_data_vector(column_count);
+
+	for (oid_t i = 0; i < column_count; i++) {
```
Consider changing your iteration variables to match Peloton's style. Instead of "i" and "to", you could use "col_iter" and "row_iter"

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+	if (!is_dict_encoded) return nullptr;
+  LOG_INFO("dictionary decode, database_id: %d, table_id: %d, tile_group_id: %d"
+					", tile_id: %d", database_id, table_id, tile_group_id, tile_id);
+  // new data mean decoded data
```
Why not call it "decoded_data" instead of "new_data_vector" then?

mbutrovich commented on test/storage/tile_compression_test.cpp:
```diff
@@ -0,0 +1,181 @@
+
+
+}
+
```
Do you ever test the GetValue and GetValueFast functions for DictEncodedTile?

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+  auto *new_tile = new Tile(backend_type, tile_group_header, original_schema, tile_group, num_tuple_slots);
+  for (oid_t i = 0; i < column_count; i++) {
+		for (oid_t to = 0; to < num_tuple_slots; to++) {
+			new_tile->SetValueFast(new_data_vector[i][to], to, original_schema.GetOffset(i), original_schema.IsInlined(i),
```
Wouldn't it be faster and use less memory to set the decoded Tile's values directly from the encoded tile's values? Why incur the cost of allocating, reading, and writing, to the separate "new_data_vector"?

mbutrovich commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+//}
+
+void DictEncodedTile::DictEncode(Tile *tile) {
+	if (is_dict_encoded) return;
```
PELOTON_ASSERT that tile isn't a nullptr.

mbutrovich commented on src/include/storage/dictionary_encoding_tile.h:
```diff
@@ -0,0 +1,148 @@
+
+ protected:
+  // is dictionary encoded
+  bool is_dict_encoded;
```
Consider trailing underscore for class data members per Google C++ Style Guide. Emailing the Peloton dev group to see if this should be enforced going forward.

mbutrovich commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,186 @@
+	for (oid_t i = 0; i < column_count; i++) {
+		// if it is inlined, no need to compress!
+		if (dict_encoded_columns.count(i) > 0) {
+			LOG_INFO("encoding column %s", schema.GetColumn(i).column_name.c_str());
```
LOG_DEBUG instead?

apavlo commented:
Closing. Resubmit PR at the end of the semester.

@dqrs @mbutrovich Nice work.

suafeng commented:
Now we support queries on dictionary encoded data. There are two ways to do it.
1. Decode the data when we try to access the encoded data, and it is our 100% goal. The implementation is that, when a data table wants to access a tile group, if this tile group is encoded (contains encoded tiles), then decode this tile group and return it.
2. The second way is to query the encoded data directly, and this is our 125% goal. The idea is that, when the database accesses the encoded column, it will see a index indicating the actual content. We moved the index to string mapping to heap, and we use this index and some pointer arithmetic to access the actual value of that column.

One thing needs to notice that in current code we are using the second way. The code of first way is commented out (mainly in src/storage/data_table.cpp). 

Modified file:
src/include/codegen/proxy/runtime_functions_proxy.h
src/codegen/proxy/runtime_functions_proxy.cpp
src/include/codegen/runtime_functions.h
src/codegen/runtime_functions.cpp
src/include/codegen/tile_group.h
src/codegen/tile_group.cpp

mbutrovich commented on src/storage/tile_group.cpp:
```diff
@@ -406,5 +408,55 @@ const std::string TileGroup::GetInfo() const {
+
+void TileGroup::DictDecode() {
+	size_t num_tile = GetTileCount();
+	for (oid_t i = 0; i < num_tile; i++) {
```
tile_idx instead of i

mbutrovich commented on src/storage/tile_group.cpp:
```diff
@@ -406,5 +408,55 @@ const std::string TileGroup::GetInfo() const {
+	ret->database_id = database_id;
+	ret->table_id = table_id;
+	ret->tile_group_id = tile_group_id;
+	for (oid_t i = 0; i < num_tile; i++) {
```
same as above. tile_idx instead of i

dqrs commented on src/storage/dictionary_encoding_tile.cpp:
```diff
@@ -0,0 +1,203 @@
+
```
Add file identity block

dqrs commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+#include <memory>
```
Add file identity block

mbutrovich commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+
+namespace peloton {
+namespace test {
+class CompressionSelectTest : public PelotonTest {};
```
Rename to TileCompressionSQLTests and rename file to tile_compression_sql_test.cpp to comply with other SQL testing conventions.

dqrs commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+  catalog::Catalog::GetInstance()->CreateDatabase(DEFAULT_DB_NAME, txn);
+  catalog::Catalog::GetInstance()->Bootstrap();
+  txn_manager.CommitTransaction(txn);
+  // Create a t
```
Fix comment

dqrs commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+	LOG_INFO("insert finish");
+	txn = txn_manager.BeginTransaction();
+	auto dataTable_object = catalog::Catalog::GetInstance()->GetTableWithName(DEFAULT_DB_NAME, DEFUALT_SCHEMA_NAME, testTableName, txn);
+	auto TGiterator = new storage::TileGroupIterator(dataTable_object);
```
Why was this allocated on the heap? Just use a local var since its deleted 10 lines later...

dqrs commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+  std::vector<FieldInfo> tuple_descriptor;
+  std::string error_message;
+
+  std::ostringstream os;
```
Bad variable name. What is os?

dqrs commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+  txn = txn_manager.BeginTransaction();
+
+  dataTable_object = catalog::Catalog::GetInstance()->GetTableWithName(DEFAULT_DB_NAME, DEFUALT_SCHEMA_NAME, testTableName, txn);
+  TGiterator = new storage::TileGroupIterator(dataTable_object);
```
Again, no need to allocate this on the heap.

mbutrovich commented on src/include/storage/dictionary_encoding_tile.h:
```diff
@@ -0,0 +1,137 @@
+/**
+ * Represents a Tile.
+ *
+ * Tiles are only instantiated via TileFactory.
```
Old comments from tile.h.

dqrs commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+  while (TGiterator->HasNext()) {
+		TGiterator->Next(tg);
+		oid_t num_tile = tg->NumTiles();
+		for (oid_t to = 0; to < num_tile; to++) {
```
Bad variable name, to. Use something more descriptive.

dqrs commented on test/storage/tile_compression_select_test.cpp:
```diff
@@ -0,0 +1,80 @@
+		oid_t num_tile = tg->NumTiles();
+		for (oid_t to = 0; to < num_tile; to++) {
+			auto curr_tile = tg->GetTileReference(to);
+			LOG_DEBUG("curr_tile id: %d, isEncoded: %d", curr_tile->GetTileId(), curr_tile->IsDictEncoded());
```
What are the expected values here? Why are you logging? These should be changed to EXPECT()s.

mbutrovich commented on test/storage/tile_compression_test.cpp:
```diff
@@ -0,0 +1,181 @@
+
+class TileCompressionTests : public PelotonTest {};
+
+TEST_F(TileCompressionTests, BasicTest) {
```
I'm failing this test:
Assertion failed: ((column_offset < schema.GetLength())), function SetValueFast, file /Users/matt/git/dictencode/src/storage/tile.cpp, line 180.

mbutrovich commented on src/include/storage/dictionary_encoding_tile.h:
```diff
@@ -0,0 +1,137 @@
+
+ protected:
+	// the string-idx mapping
+  std::unordered_map<type::Value, uint8_t, type::Value::hash, type::Value::compress_equal_to> dict;
```
Why uint8_t for value if you're reinterpret_casting a uint32_t on insertion and delete? dictionary_encoding_tile.cpp lines 129-132.

apavlo commented:
We have decided not to merge this project.
